apm-agent-core/src/main/java/co/elastic/apm/agent/util/VersionUtils.java

```diff
@@ -43,7 +47,40 @@ public static String getAgentVersion() {
     }
 
     @Nullable
-    public static String getVersionFromPomProperties(Class clazz, String groupId, String artifactId) {
+    public static String getVersion(Class<?> clazz, String groupId, String artifactId) {
+        String version = getFromCache(clazz);
+        if (version != null) {
+            return version;
+        }
+        version = getVersionFromPackage(clazz);
+        if (version == null) {
+            version = getVersionFromPomProperties(clazz, groupId, artifactId);
+        }
+        cacheValue(clazz, version);
+        return version;
+    }
+
+    @Nullable
+    static String getVersionFromPackage(Class<?> clazz) {
+        Package pkg = clazz.getPackage();
+        if (pkg != null) {
+            return pkg.getImplementationVersion();
+        }
+        return null;
+    }
+
+    @Nullable
+    private static String getFromCache(Class<?> clazz) {
+        String cached = versionsCache.get(clazz);
+        return cached != UNKNOWN_VERSION ? cached : null;
```

Comment:
This means it is effectively not caching the result of failing to get the version

Reply:
oh boy, good point